src/log/builder.cr

```diff
@@ -0,0 +1,114 @@
+require "weak_ref"
+
+# A `Log::Builder` creates `Log` instances for a given source.
+# It allows you to bind sources and patterns to a given backend.
+# Already created `Log` will be reconfigured as needed.
+class Log::Builder
+  @mutex = Mutex.new(:unchecked)
+  @logs = Hash(String, WeakRef(Log)).new
+
+  private record Binding, source : String, level : Severity, backend : Backend
+  @bindings = Array(Binding).new
+
+  # Binds a *source* pattern to a *backend* for all logs that are of severity equal or higher to *level*.
+  def bind(source : String, level : Severity, backend : Backend)
+    # TODO validate source is a valid path
+
+    @mutex.synchronize do
+      binding = Binding.new(source: source, level: level, backend: backend)
+      @bindings << binding
+
+      each_log do |log|
+        if Builder.matches(log.source, binding.source)
+          append_backend(log, binding.level, binding.backend)
+        end
+      end
+    end
+  end
+
+  # Removes all existing bindings.
+  def clear
+    @mutex.synchronize do
+      @bindings.clear
+      each_log do |log|
+        log.backend = nil
+        log.initial_level = :none
+      end
+    end
+  end
+
+  # Returns a `Log` for the given *source* with a severity level and
+  # backend according to the bindings in *self*.
+  # If new bindings are applied, the existing `Log` instances will be
+  # reconfigured.
+  # Calling this method multiple times with the same value will return
+  # the same object.
+  def for(source : String) : Log
+    @mutex.synchronize do
+      log = @logs[source]?.try &.value
+
+      if log.nil?
+        log = Log.new(source, nil, :none)
+        @bindings.each do |binding|
+          next unless Builder.matches(log.source, binding.source)
+          append_backend(log, binding.level, binding.backend)
+        end
+        @logs[source] = WeakRef.new(log)
+      end
+
+      log
+    end
+  end
+
+  # :nodoc:
+  private def each_log
+    @logs.reject! { |_, log_ref| log_ref.value.nil? }
```

Comment:
This is unexpected side-effect.

Reply:
I don’t follow

Reply:
Why calling  `each_log` mutates `@logs`?

Reply:
I'm almost sure it's because they are weak refs. So this is returning only the valid ones, while making sure the invalid ones are deleted. I think this fine.

Reply:
@asterite I'm not questioning the logic, but the side-effect in itself - `each` (and friends) is suppose to enumerate the values, not modify 'em... it could be easily done without modifying `@logs` and if that's intended there should be a better place to do it, not in here...

Reply:
You mean like moving this line to a helper function?

The idea of this function is to iterate the log entries. Right before doing that it makes sure to remove invalid entries. Why is that wrong?

Reply:
Would you move that line or a helper function to each call site?

Reply:
> Would you move that line or a helper function to each call site?

Moving that into a helper method would IMO be a cleaner design, yes. _each call site_ counts at exactly 2 occurrences, so it's not like that would be adding a ton of overhead anyway.

Reply:
The idea is for call sites to not have to know all of this.

I guess it's double checked because weak refs can become nil at any point.

Reply:
@Sija I don't think it matters. Whether `@logs` contains deleted references or not is only internal information. To the outside both states count as non-existing.

Reply:
Fair 'nuff, thanks for bearing with me.

Reply:
One last question: is it needed at all, given the check below?